massa-network-worker/src/network_worker.rs

```diff
@@ -96,55 +109,119 @@ impl NetworkWorker {
 
-        // wake up the controller at a regular interval to retry connections
-        let mut wakeup_interval = tokio::time::interval(self.cfg.wakeup_interval.to_duration());
-        let mut need_connect_retry = true;
+        // Start a task to run the listener.
+        // TODO: config size.
+        let (listener_tx, listener_rx) = bounded::<(ReadHalf, WriteHalf, SocketAddr)>(1);
+        let mut listener = self
+            .listener
+            .take()
+            .expect("No listener at start of run loop.");
+        let listener_handle = self.runtime.spawn(async move {
+            loop {
+                match listener.accept().await {
+                    Ok(res) => {
+                        let sender = listener_tx.clone();
+                        Handle::current()
+                            .spawn_blocking(move || {
+                                sender
+                                    .send(res)
+                                    .expect("Failed to send listener message to network worker.");
+                            })
+                            .await
+                            .expect(
+                                "Failed to run task to send listener message to network worker.",
+                            );
+                    }
+                    Err(err) => {
+                        debug!("connection accept failed: {}", err);
+                        massa_trace!("in_connection_failed", {"err": err.to_string()});
+                        break;
+                    }
+                }
+            }
+        });
 
         loop {
-            if need_connect_retry {
+            if out_connection_rx.is_none() {
+                // Scope the channel creation to here,
+                // to ensure all senders drop when the futureset is empty.
+                let (out_connection_tx, rx) = bounded::<(
+                    tokio::io::Result<(ReadHalf, WriteHalf)>,
+                    IpAddr,
+                )>(self.cfg.node_command_channel_size); // TODO: config
+
+                out_connection_rx = Some(rx);
+
                 // try to connect to candidate IPs
+                let mut out_connecting_futures = FuturesUnordered::new();
                 let candidate_ips = self.peer_info_db.get_out_connection_candidate_ips()?;
                 for ip in candidate_ips {
                     debug!("starting outgoing connection attempt towards ip={}", ip);
                     massa_trace!("out_connection_attempt_start", { "ip": ip });
                     self.peer_info_db.new_out_connection_attempt(&ip)?;
-                    let mut connector = self
-                        .establisher
-                        .get_connector(self.cfg.connect_timeout)
-                        .await?;
+
+                    // Run the future on the current thread.
+                    // Note: this future doesn't actually await anything, consider removing async.
```

Comment:
I think this can be done in this PR regarding the code in `get_connector` which is pretty straight forward, it will simplify the readability here.

Reply:
I believe I tried and compilation failed in places that were too numerous to tackle in this PR.